tests/test_etoro.py

```diff
@@ -0,0 +1,224 @@
+# This file is part of Maker Keeper Framework.
+#
+# Copyright (C) 2019 MikeHathaway 
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import json
+import os
+import re
+import time
+
+from pymaker import Wad
+from pyexchange.etoro import EToroApi, Order, Trade
+
+# Models HTTP response, produced by EToroMockServer
+class MockedResponse:
+    def __init__(self, text: str, status_code=200):
+        assert (isinstance(text, str))
+        assert (isinstance(status_code, int))
+        self.status_code = status_code
+        self.ok = 200 <= status_code < 400
+        self.text = text
+        self.reason = None
+
+    def json(self, **kwargs):
+        return json.loads(self.text)
+
+# Determines response to provide based on the requested URL
+class EToroMockServer:
+    # Read JSON responses from a pipe-delimited file, avoiding JSON-inside-JSON parsing complexities
+    responses = {}
+    cwd = os.path.dirname(os.path.realpath(__file__))
+    response_file_path = os.path.join(cwd, "mock/etoro-api-responses")
+    with open(response_file_path, 'r') as file:
+        for line in file:
+            kvp = line.split("|")
+            assert(len(kvp) == 2)
+            responses[kvp[0]] = kvp[1]
+
+    @staticmethod
+    def handle_request(**kwargs):
+        assert("url" in kwargs)
+        url = kwargs["url"]
+        method = kwargs["method"]
+        if method == "GET":
+            return EToroMockServer.handle_get(url)
+        elif method == "POST":
+            return EToroMockServer.handle_post(url, kwargs["data"])
+        else:
+            return EToroMockServer.handle_delete(url)
+
+    @staticmethod
+    def handle_get(url: str):
+        # Parse the URL to determine which piece of canned data to return
+        if re.search(r"api\/v1\/instruments", url):
+            return MockedResponse(text=EToroMockServer.responses["markets"])
+        elif re.search(r"api\/v1\/balances", url):
+            return MockedResponse(text=EToroMockServer.responses["balances"])
+        elif re.search(r"api\/v1\/orders", url):
+            return MockedResponse(text=EToroMockServer.responses["orders"])
+        elif re.search(r"\/api\/v1\/trades", url):
+            return MockedResponse(text=EToroMockServer.responses["trades"])
+        else:
+            raise Exception("Unable to match HTTP GET request to canned response")
```

Comment:
Although you'll find it all over our code, I recommend not throwing the base `Exception` class.  Makes it harder for consumers to catch anticipated exceptions and avoid catching ones which shouldn't be caught.

Reply:
Updated to ValueError and added the test URL and Data field as part of the message